src/BizHawk.Client.Common/zstd/Zstd.cs

```diff
@@ -0,0 +1,342 @@
+﻿using System;
+using System.IO;
+using System.Runtime.InteropServices;
+
+using BizHawk.BizInvoke;
+using BizHawk.Common;
+
+namespace BizHawk.Client.Common
+{
+	public static class Zstd
+	{
+		private class ZstdCompressionStream : Stream
+		{
+			private readonly Stream _baseStream;
+			private readonly IntPtr _zcs;
+
+			private readonly byte[] _inputBuffer;
+			private readonly GCHandle _inputHandle;
+			private LibZstd.StreamBuffer _input;
+
+			private readonly byte[] _outputBuffer;
+			private readonly GCHandle _outputHandle;
+			private LibZstd.StreamBuffer _output;
+
+			// TODO: tweak these sizes
+
+			// 4 MB input buffer
+			private const int INPUT_BUFFER_SIZE = 1024 * 1024 * 4;
+			// 1 MB output buffer
+			private const int OUTPUT_BUFFER_SIZE = 1024 * 1024 * 1;

```

Comment:
Would recommend using `ZSTD_CStreamInSize` and `ZSTD_CStreamOutSize` here instead of these hardcoded (and very large) sizes.

Reply:
I used very large sizes due to the .h file stated those functions are only good in a C environment, in an environment like C# where there is a cost to interop it's recommended to just use very big buffers to reduce interop calls.

Reply:
Do you have some link for that? I can't find it right now, and when testing myself I'm also not really seeing any performance loss from using the recommended buffer sizes.

Reply:
https://github.com/facebook/zstd/blob/f5c4ec465817f77ea04346659e284e2c1c5c6f27/lib/zstd.h#L724-L731

Reply:
Thanks. I guess what I just don't like is the fact that 5MB of buffer are allocated every time a zstd stream is initialized, which could happen as often as once per frame (300mb of managed allocations per second).

Reply:
Perhaps this could be reworked to have a separate "context" class, dictionary of which is kept around statically and may be reused for different streams (only if said dictionary doesn't have an unused context instead wil a new context is created and added).

Reply:
Ideally we'd get rid of the extra buffers+copies entirely, but at least it's faster than what we have now

Reply:
See e5514e25e5611a1adc1896ed558186c29e754801